FileServ/FileServer.cpp

```diff
@@ -254,42 +215,51 @@ void cb_downloadStart(FileServer_Private& client)
         return;
     }
 
-    std::unique_ptr<FileRequest> req(new FileRequest(fd));
-    client.m_buffer.write<uint32_t>(DS::e_NetSuccess);
-    client.m_buffer.write<uint32_t>(++client.m_readerId);
-    client.m_buffer.write<uint32_t>(req->m_size);
-
-    FileRequest::chunk_t c = req->nextChunk();
-    client.m_buffer.write<uint32_t>(std::get<1>(c));
-    SEND_FD_REPLY(req->m_fd, req->m_pos, std::get<1>(c));
-    if (!std::get<0>(c)) {
-        client.m_downloads[client.m_readerId] = std::move(req);
-    }
-}
-
-void cb_downloadNext(FileServer_Private& client)
-{
-    START_REPLY(e_FileToCli_FileDownloadReply);
+    struct stat stat_buf;
+    off_t filepos = 0;
+    if (fstat(fd, &stat_buf) < 0) {
+        ST::printf(stderr, "[File] Could not stat file {}\n[File] Requested by {}\n",
+                   filename, DS::SockIpAddress(client.m_sock));
+        START_REPLY(e_FileToCli_FileDownloadReply);
+        client.m_buffer.write<uint32_t>(transId);
+        client.m_buffer.write<uint32_t>(DS::e_NetFileNotFound);
+        client.m_buffer.write<uint32_t>(0);     // Reader ID
+        client.m_buffer.write<uint32_t>(0);     // File size
+        client.m_buffer.write<uint32_t>(0);     // Data packet size
+        SEND_REPLY();
 
-    uint32_t transId = DS::RecvValue<uint32_t>(client.m_sock);
-    uint32_t readerId = DS::RecvValue<uint32_t>(client.m_sock);
-    auto fi = client.m_downloads.find(readerId);
-    if (fi == client.m_downloads.end()) {
-        // The last chunk was already sent, we don't care anymore
+        close(fd);
         return;
     }
 
-    client.m_buffer.write<uint32_t>(transId);
-    client.m_buffer.write<uint32_t>(DS::e_NetSuccess);
-    client.m_buffer.write<uint32_t>(fi->first);
-    client.m_buffer.write<uint32_t>(fi->second->m_size);
-
-    FileRequest::chunk_t c = fi->second->nextChunk();
-    client.m_buffer.write<uint32_t>(std::get<1>(c));
-    SEND_FD_REPLY(fi->second->m_fd, fi->second->m_pos, std::get<1>(c));
-    if (std::get<0>(c)) {
-        client.m_downloads.erase(fi);
+    // Attempting to use the MOUL protocol's "acking" of file chunks has a severe performance
+    // penalty. For me, I see an improvement from 950 KiB/s to 14 MiB/s by simply sending the
+    // whole file synchronously. We still have to chunk it, though, so the client's progress
+    // bar updates correctly. Besides, this is TCP, who needs acks at this level? The drawback
+    // to this is that the connection becomes unresponsive to all other traffic when a download
+    // is in progress.
```

Comment:
One of my eventual desires is to port DS's lazy threaded-blocking network stack to ASIO as well, which would solve this issue...

Reply:
A lazy workaround would be to add in a secondary loop here using `poll()` that will process any request except for another download. But, that seems a solution looking for a problem atm.